internal/manager/config.go

```diff
@@ -185,16 +187,77 @@ func (c *Config) FlagSet() *pflag.FlagSet {
 	return flagSet
 }
 
+// CheckKongNonDefaultWorkSpace invokes :workspaces/ws-name, checks workspace existence
+func (c *Config) CheckKongNonDefaultWorkSpace(ctx context.Context, adminURL string, wsName string, client *kong.Client) (bool, error) {
+	exists, err := client.Workspaces.Exists(ctx, kong.String(wsName))
+	if err != nil {
+		return false, fmt.Errorf("failed looking up workspace: %w", err)
+	}
+
+	// if the provided workspace does not exist, for convenience we create it.
+	if !exists {
+		workspace := kong.Workspace{
+			Name: kong.String(wsName),
+		}
+		_, err := client.Workspaces.Create(ctx, &workspace)
+		if err != nil {
+			return false, fmt.Errorf("failed creating workspace: %w", err)
+		}
+		fmt.Printf("created workspace %s successfully", wsName)
+	}
+	return true, nil
+}
+
+// GetKongClient create a kong client based on configuration
 func (c *Config) GetKongClient(ctx context.Context) (*kong.Client, error) {
 	if c.KongAdminToken != "" {
+		fmt.Printf("kong-admin-token %s", c.KongAdminToken)
 		c.KongAdminAPIConfig.Headers = append(c.KongAdminAPIConfig.Headers, "kong-admin-token:"+c.KongAdminToken)
 	}
+
 	httpclient, err := adminapi.MakeHTTPClient(&c.KongAdminAPIConfig)
 	if err != nil {
 		return nil, err
 	}
 
-	return adminapi.GetKongClientForWorkspace(ctx, c.KongAdminURL, c.KongWorkspace, httpclient)
+	client, err := kong.NewClient(kong.String(c.KongAdminURL), httpclient)
+	if err != nil {
+		return nil, fmt.Errorf("creating Kong client err: %w", err)
+	}
+
+	if c.KongWorkspace == "" {
+		return client, nil
+	}
+
+	exists, err := c.CheckKongNonDefaultWorkSpace(ctx, c.KongAdminURL, c.KongWorkspace, client)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, fmt.Errorf("workspace: %s does not exist", c.KongWorkspace)
+	}
+
+	if c.KongNonAdminToken == "" {
+		// set up kong-admin user with workspace
+		client.SetWorkspace(c.KongWorkspace)
+		return client, nil
+	}
+
+	c.KongAdminAPIConfig.Headers = []string{}
+	c.KongAdminAPIConfig.Headers = append(c.KongAdminAPIConfig.Headers, "kong-admin-token:"+c.KongNonAdminToken)
+	httpclient, err = adminapi.MakeHTTPClient(&c.KongAdminAPIConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err = kong.NewClient(kong.String(c.KongAdminURL), httpclient)
+	if err != nil {
+		return nil, fmt.Errorf("creating Kong client: %w", err)
+	}
+
+	client.SetWorkspace(c.KongWorkspace)
+	return client, nil
```

Comment:
question: It's not clear to me why we moved this function from it's previous location, can we just use this function optionally as needed?

Reply:
The justification is https://github.com/Kong/kubernetes-ingress-controller/pull/1758#discussion_r710590012

Reply:
5ac1d90c5e6aa8319c4fd3169d6e45da4810e7cf